_posts/2021-07-30-iceberg-concurrency-snapshots-spec.md

```diff
@@ -0,0 +1,327 @@
+);
+```
+
+Let’s check the snapshot table again:
+
+```
+SELECT snapshot_id, parent_id, operation
+FROM iceberg.logging.“events$snapshots”;
+```
+
+Result:
+
+| snapshot_id | parent_id | operation |
+| --- | --- | --- |
+| 7620328658793169607 | | append |
+| 2115743741823353537 | 7620328658793169607 | append |
+| 7030511368881343137 | 2115743741823353537 | append |
+
+Let’s also verify that our row was added:
+
+```
+SELECT level, message
+FROM iceberg.logging.events;
+```
+
+Result:
+
+| level | message |
+| --- | --- |
+|ERROR|Oh noes |
+|INFO |It is all good |
+|ERROR|Double oh noes |
+|WARN |Maybeh oh noes?|
+
+Since Iceberg is already tracking the list of files added and removed at each
+snapshot, it would make sense that you can travel back and forth between these
+different views into the system, right? This concept is called time traveling.
+You need to specify which snapshot you would like to read from and you will see
+the view of the data at that timestamp. In Trino, you need to use the `@`
+operator, followed by the snapshot you wish to read from:
+
+```
+SELECT level, message
+FROM iceberg.logging.“events@2115743741823353537”;
+```
+
+Result:
+
+| level | message |
+| --- | --- |
+|ERROR|Double oh noes |
+|WARN |Maybeh oh noes?|
+|ERROR|Oh noes |
+
+If you determine there is some issue with your data, you can always roll back to
+the previous state permanently as well. In Trino we have a function called
+`rollback_to_snapshot` to move the table state to another snapshot:
+
+```
+CALL system.rollback_to_snapshot(‘logging’, ‘events’, 2115743741823353537);
+```
+
+Now that we have rolled back, observe what happens when we query the events
+table with:
+
+```
+SELECT level, message
+FROM iceberg.logging.events;
+```
+
+Result:
+
+| level | message |
+| --- | --- |
+|ERROR|Double oh noes |
+|WARN |Maybeh oh noes?|
+|ERROR|Oh noes |
+
+Notice the `INFO` row is still missing even though we query the table without
+specifying a snapshot id. Now just because we rolled back, doesn’t mean we’ve
+lost the snapshot we just rolled back from. In fact, we can roll forward, or as
+I like to call it, 
+[back to the future](https://en.wikipedia.org/wiki/Back_to_the_Future)! In
+Trino, you use the same function call but with a predecessor of the existing
+snapshot:
+
+```
+CALL system.rollback_to_snapshot(‘logging’, ‘events’, 7030511368881343137)
+```
+
+And now we should be able to query the table again and see the `INFO` row 
+return:
+
+```
+SELECT level, message
+FROM iceberg.logging.events;
+```
+
+Result:
+
+| level | message |
+| --- | --- |
+|ERROR|Oh noes |
+|INFO |It is all good |
+|ERROR|Double oh noes |
+|WARN |Maybeh oh noes?|
+
+As expected, the INFO row returns when you roll back to the future.
+
+Having snapshots not only provides you with a level of immutability that is key
+to the eventual consistency model, but gives you a rich set of features to
+version and move between different versions of your data like a git repository.
```

Comment:
maybe add a link to the newer and better syntax for time travel

Reply:
https://github.com/trinodb/trino/issues/12745

That's still in the works AFAICT